### PR TITLE
Remove swarm inspect and use info instead

### LINKED ISSUE #24148: `docker swarm inspect` and `docker info` inconsistencies
Both `docker info` and `docker swarm inspect` display information regarding the swarm.

They do so in a different way.

There are also other inconsistencies in `swarm inspect` (#23689, #24122).

/cc @stevvooe @aanand @dnephin @tonistiigi

### PR DESCRIPTION
**Pull request by vdemeester**
Remove the `docker swarm inspect` command and use `docker info` instead to display swarm information if the current node is a manager 🐮.

This only change the cli side and the return of `/info` but not the `/swarm/inspect` endpoint. The _attribute_ used in the info endpoint is `Swarm.Cluster` and is populated only when the node is a manager.

```
$ docker info
# […]
Swarm: active
 NodeID: b1n3yqwdmvgpbhgmiv4xh48xy
 Is Manager: true
 ClusterID: d3aen0t7fkdw6kq1y4e3w4vv6
 Managers: 1
 Nodes: 1
 Name: default
 Orchestration:
  Task History Retention: 10
 Raft:
  Snapshot interval: 10000
  Heartbeat tick: 1
  Election tick: 3
 Dispatcher:
  Heartbeat period: 5 seconds
 CA configuration:
  Expiry duration: 3 months
# […]
```

I did not add all the thing displayed by `swarm inspect` but they are available through the API.
- [x] Complete attribute displayed ?
- [x] <del>Remove `/swarm/inspect` endpoint ?</del>

I'll create a `engine-api` if this pass design-review :wink: 

Closes #24148

/cc @aluzzardi @tiborvass @stevvooe @tonistiigi 

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

File: api/client/system/info.go [hunk 1]
```diff
@@ -87,6 +87,26 @@ func runInfo(dockerCli *client.DockerCli) error {
...
 			fmt.Fprintf(dockerCli.Out(), " Nodes: %d\n", info.Swarm.Nodes)
 			ioutils.FprintfIfNotEmpty(dockerCli.Out(), " CACertHash: %s\n", info.Swarm.CACertHash)
+			fmt.Fprintf(dockerCli.Out(), "Cluster:\n")
+			fmt.Fprintf(dockerCli.Out(), " ID: %s\n", info.Swarm.Cluster.ID)
+			fmt.Fprintf(dockerCli.Out(), " Version: %d\n", info.Swarm.Cluster.Meta.Version.Index)
+			fmt.Fprintf(dockerCli.Out(), " Name: %s\n", info.Swarm.Cluster.Spec.Annotations.Name)
+			fmt.Fprintf(dockerCli.Out(), " Acceptance Policies:\n")
+			for _, policy := range info.Swarm.Cluster.Spec.AcceptancePolicy.Policies {
+				fmt.Fprintf(dockerCli.Out(), "  %s:\n", policy.Role)
+				fmt.Fprintf(dockerCli.Out(), "   auto-accept: %t\n", policy.Autoaccept)
```
Comment (AkihiroSuda) [inline api/client/system/info.go]: I feel `Auto-acceptance` and `Secret` are better for consistency

Comment (vdemeester) [inline api/client/system/info.go]: @AkihiroSuda you're right :+1:

**Comment by icecrime** [comment]
A few remarks about acceptance policies:
- I don't think we should display the secret: people copy/paste this in GitHub issues :innocent: 
- We can probably condense the output (e.g., `Auto-accept: (manager|worker|none)`).

**Comment by vdemeester** [comment]
Rebased and hopefully, tests fixed 👼

@icecrime 

> I don't think we should display the secret: people copy/paste this in GitHub issues :innocent: 

`docker swarm inspect` did display it but I'm ok with not displaying in `docker info` (it will still be present in the json)

> We can probably condense the output (e.g., `Auto-accept: (manager|worker|none)`).

That's an idea, I'll look into it.

**Comment by vdemeester** [comment]
Hum, removing `/swarm/inspect`means removing `client.SwarmInspect` (from `engine-api`) and means making `swarm update` call the `/info` endpoint instead. 

I'm not sure we should remove it, just adding *Cluster* information in `/info`.

@thaJeztah @icecrime wdyt ?

**Comment by thaJeztah** [comment]
Perhaps we can keep the  `inspect` endpoint for now, if that simplifies things

**Comment by tiborvass** [comment]
@vdemeester sorry but would you mind linking where `swarm update` calls `/swarm` (inspect) ?

**Comment by vdemeester** [comment]
@tiborvass yes, here https://github.com/docker/docker/blob/master/api/client/swarm/update.go#L35 👼

**Comment by icecrime** [comment]
I think it's ok for `swarm update` to call the `/info` endpoint as long as the `SwarmSpec` is a nested element of the result, and not something that I need to construct myself for the purpose of calling `/swarm/update`.

**Comment by icecrime** [comment]
Also :wink: 

```
11:22:56 FAIL: docker_cli_swarm_test.go:15: DockerSwarmSuite.TestSwarmUpdate
11:22:56 
11:22:56 [d66998809] waiting for daemon to start
11:22:56 [d66998809] daemon started
11:22:56 docker_cli_swarm_test.go:24:
11:22:56     c.Assert(err, checker.IsNil, check.Commentf("out: %v", out))
11:22:56 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc822f77080), Stderr:[]uint8(nil)} ("exit status 1")
11:22:56 ... out: Error response from daemon: page not found
```

**Comment by vdemeester** [comment]
@icecrime ok, I'll update and prepare a PR for engine-api. The test failure is related to that :wink:

**Comment by vdemeester** [comment]
Updated, fixed test failures and created a PR in `engine-api` (https://github.com/docker/engine-api/issues/318) — vendor is expected to fail 👼.

**Comment by tiborvass** [comment]
@vdemeester let's keep the API unchanged and have `inspect` make multiple calls for now. Not sure if you would still need #318 then.

**Comment by vdemeester** [comment]
@tiborvass you mean `info` 👼

**Comment by vdemeester** [comment]
Updated 👼

**Comment by vieux** [comment]
@vdemeester `docker_cli_swarm_test.go` uses `docker swarm inspect`

**Comment by vdemeester** [comment]
@vieux ah damn.. revert the wrong file… I'll fix 👼

**Comment by tiborvass** [comment]
@vdemeester some updates we need on this one:

- Rebase :)
- Update description to show the new information, since the result from the swarm API has changed (no more acceptance policy, CAhash, etc.)
- Make sure we DO NOT display join tokens (because they are returned by the API)
- Make only 1 API call to /info, and gather all the swarm information server side.
- Having `Cluster:` and `Swarm:` makes no sense, they are synonyms, they should be merged, and `ID` be renamed to `ClusterID`.
- Remove `Version` from `Cluster:`
- `Heartbeat period` should use human formatting for duration

**Comment by kencochrane** [comment]
Can we add an `IsLeader` field to show if this node is the lead manager. I'm currently using swarm inspect for this, but if swarm inspect is going away, then I need another way to get that value.

Another question. Will `Cluster:Id` be available on the worker nodes as well as the manager nodes? It isn't today and that would be very helpful. It is an easy way to see what swarm a node is in for logging and analytics/metrics gathering for nodes by cluster.

**Comment by vdemeester** [comment]
> Make only 1 API call to /info, and gather all the swarm information server side.

Ok this mean I need to make a PR in engine-api 👼

**Comment by aluzzardi** [comment]
Hey Ken, regarding IsLeader, I think you can grab that from docker node
inspect self -f "{{ something something }}"

I do agree it'd be useful in docker info regardless so +1 for that request

**Comment by kencochrane** [comment]
@aluzzardi thanks for the work around. The one issue with `docker node` is that it only works for manager nodes, where as docker info works for all mode types.

**Comment by aluzzardi** [comment]
Indeed - but for your use case, you can assume that if it's not a manager
node then it's probably not the leader :)

However I do still agree with you that this information should be available
on docker info - just pointing out a workaround

**Comment by tiborvass** [comment]
@vdemeester I was thinking about not returning join tokens via the /info API at all not just CLI

**Comment by tiborvass** [comment]
Carried in #25042

**Pull request closed without merging**